File: DatabaseTypes/Report.fs
namespace DatabaseTypes

open System
open System.Text.Json.Serialization
open Utils
open DatabaseTypes.Identificators
open Utils.Converters

[<JsonConverter(typeof<ReportTypeConverter>)>]
type ReportType private () =
    inherit BaseType("report")
    static member Instance = ReportType()
and ReportTypeConverter() =
    inherit StringConverter<ReportType>((fun m -> m.Value), (fun _ -> ReportType.Instance))

type ProblemReport = {
    [<JsonPropertyName("generated_problem_id")>]
    GeneratedProblemId : GeneratedProblemId
    [<JsonPropertyName("answer")>]
    [<JsonConverter(typeof<OptionValueConverter<string>>)>]
    Answer : string option
    [<JsonPropertyName("expected_answer")>]
    ExpectedAnswer : string
    [<JsonPropertyName("is_correct")>]
    IsCorrect : bool
    [<JsonPropertyName("timestamp")>]
    [<JsonConverter(typeof<OptionValueConverter<DateTimeOffset>>)>]
    Timestamp : DateTimeOffset option
}

type Report =
    {
        [<JsonPropertyName("id")>]
        Id : ReportId
        [<JsonPropertyName("generated_problem_set_id")>]
        GeneratedProblemSetId : GeneratedProblemSetId
        [<JsonPropertyName("submission_id")>]
        SubmissionId : SubmissionId
        [<JsonPropertyName("permissions")>]
        Permissions : Permissions
        [<JsonPropertyName("started_at")>]
        StartedAt : DateTimeOffset
        [<JsonPropertyName("finished_at")>]
        FinishedAt : DateTimeOffset
        [<JsonPropertyName("answers")>]
        Answers : List<ProblemReport>
        [<JsonPropertyName("type")>]
        Type : ReportType
    }

    static member CreateDocumentKey(id: ReportId): DocumentKey =
        DocumentKey.Create(id.Value, ReportType.Instance.Value)
    member private this.DocKey = Report.CreateDocumentKey(this.Id)

    interface IDocumentKey with
        member this.Type
            with get() = this.DocKey.Type
        member this.Key
            with get() = this.DocKey.Key

File: react-ui/src/models/Report.ts
import { GeneratedProblemId, GeneratedProblemSetId, SubmissionId } from "./Identificators";
import { GeneratedView } from "./GeneratedView";
import { User } from "./User";

export interface ProblemReport {
    id: GeneratedProblemId;
    title: string;
    view: GeneratedView;
    answer?: string;
    expected_answer: string;
    is_correct: boolean;
    timestamp?: Date;
}

export interface ProblemSetReport {
    id: GeneratedProblemSetId;
    title: string;
    problems: Array<ProblemReport>;
}

export interface Report {
    id: string;
    problem_set: ProblemSetReport;
    started_at: Date;
    finished_at: Date;
    author: User;
}